### PR TITLE
feat: Improved threshold handling.

### PR DESCRIPTION
**Pull request by ShogunPanda**
This PR reworks threshold handling to make sure it works as intended.

Fixes and supersedes #417.

### CODE REVIEW

File: packages/orama/src/components/algorithms.ts [hunk 1]
```diff
@@ -41,16 +49,47 @@ export function prioritizeTokenScores(arrays: TokenScore[][], boost: number, thr
...
-    return results.slice(0, shortestArrayLength)
+    return results.slice(0, lastTokenWithAllKeywords + 1)
   }
 
   // If the threshold is between 0 and 1, we will return all the results that contains at least the threshold of search terms
   // For example, if threshold is 0.5, we will return all the results that contains at least 50% of the search terms
   // (fuzzy match with a minimum threshold)
-  const thresholdLength = Math.ceil((threshold * 100 * results.length) / 100)
+  const thresholdLength =
+    lastTokenWithAllKeywords + Math.ceil((threshold * 100 * (results.length - lastTokenWithAllKeywords)) / 100)
```
Comment (allevo) [inline packages/orama/src/components/algorithms.ts:92]: Are there some tests that test this? In particular, the comment above.

Comment (ShogunPanda) [reply packages/orama/src/components/algorithms.ts:92]: Yup, see: https://github.com/oramasearch/orama/blob/improved-sort/packages/orama/tests/threshold.test.ts#L128

File: packages/orama/src/components/algorithms.ts [hunk 2]
```diff
@@ -41,16 +49,47 @@ export function prioritizeTokenScores(arrays: TokenScore[][], boost: number, thr
...
+  const allResults = results.length
+  const tokenKeywordsCount: [InternalDocumentID, number][] = []
+
+  for (const tokenKeywordsCountEntry of tokenKeywordsCountMap.entries()) {
+    tokenKeywordsCount.push(tokenKeywordsCountEntry)
+  }
+
+  // Find the index of the last result with all keywords.
+  // Note that since score is multipled by 1.5 any time the token is encountered in results it means
+  // that tokenMap and tokenKeywords should always have the same order.
```
[the remark this answers is not in the document]
Comment (ShogunPanda) [reply packages/orama/src/components/algorithms.ts:62]: Yeah, I renamed the variable and forgot about the comment. Fixing it.

About the addition below: later we actually sort by score and number of keywords, so this doesn't really matter.
My only point here is to emphasize that if a document has a significantly bigger score than another, it means it appeared in more keywords search than the other and thus `tokenScoresMap` and `tokenKeywordsCountMap` have the same key order if we order by values.

**Review by allevo — APPROVED** [review]
LGTM

**Review by micheleriva — APPROVED** [review]
LGTM

**Pull request merged**